File: Design-Patterns-Using-Ruby/Builder/BuilderPatternHouse.rb
# A class representing a house
class House
    attr_accessor :wall, :roof
  
    def initialize(wall, roof)
        @wall = wall
        @roof = roof
    end
  
    def to_s
        "House of #{@wall} and #{@roof}"
    end
end

# An abstract builder class that specifies the interface for building a house
class HouseBuilder
    attr_accessor :house
  
    def initialize
        @house = House.new("", "")
    end
  
    def set_wall
        raise NotImplementedError, 'Subclasses must implement this method'
    end
  
    def set_roof
        raise NotImplementedError, 'Subclasses must implement this method'
    end
  
    def get_house
        temp = @house
        @house = House.new("", "") # assign new house.
        temp
    end
end

# A builder class that builds a wooden house
class WoodenHouseBuilder < HouseBuilder
    def set_wall
        @house.wall = "Wooden Wall"
        self
    end
  
    def set_roof
        @house.roof = "Wooden Roof"
        self
    end
end

# A builder class that builds a concrete house
class ConcreteHouseBuilder < HouseBuilder
    def set_wall
        @house.wall = "Concrete Wall"
        self
    end
  
    def set_roof
        @house.roof = "Concrete Roof"
        self
    end
end

# A class that directs the building of a house
class HouseDirector
    attr_accessor :builder
  
    def initialize(builder)
        @builder = builder
    end
  
    def construct
        @builder.set_wall.set_roof.get_house
    end
end

# Client code
builder = ConcreteHouseBuilder.new
director = HouseDirector.new(builder)
house = director.construct
puts house

builder = WoodenHouseBuilder.new
director = HouseDirector.new(builder)
house2 = director.construct
puts house2

=begin 
House of Concrete Wall and Concrete Roof
House of Wooden Wall and Wooden Roof
=end

File: Design-Patterns-Using-Swift/Builder/BuilderPatternHouse.swift
// A class representing a house
class House {
    private var wall: String
    private var roof: String

    init(wall: String, roof: String) {
        self.wall = wall
        self.roof = roof
    }

    func setWall(_ wall: String) {
        self.wall = wall
    }

    func setRoof(_ roof: String) {
        self.roof = roof
    }

    var description: String {
        return "House of \(wall) and \(roof)"
    }
}

// An abstract builder class that specifies the interface for building a house
class HouseBuilder {
    var house: House

    init() {
        self.house = House(wall: "", roof: "")
    }

    func setWall() -> Self {
        fatalError("Abstract method. Subclasses must implement this method.")
    }

    func setRoof() -> Self {
        fatalError("Abstract method. Subclasses must implement this method.")
    }

    func getHouse() -> House {
        let temp = self.house
        self.house = House(wall: "", roof: "") // assign new house.
        return temp
    }
}

// A builder class that builds a wooden house
class WoodenHouseBuilder: HouseBuilder {
    override func setWall() -> Self {
        self.house.setWall("Wooden Wall")
        return self
    }

    override func setRoof() -> Self {
        self.house.setRoof("Wooden Roof")
        return self
    }
}

// A builder class that builds a concrete house
class ConcreteHouseBuilder: HouseBuilder {
    override func setWall() -> Self {
        self.house.setWall("Concrete Wall")
        return self
    }

    override func setRoof() -> Self {
        self.house.setRoof("Concrete Roof")
        return self
    }
}

// A class that directs the building of a house
class HouseDirector {
    private var builder: HouseBuilder

    init(builder: HouseBuilder) {
        self.builder = builder
    }

    func construct() -> House {
        return self.builder.setWall().setRoof().getHouse()
    }
}

// Client code
let builder = ConcreteHouseBuilder()
let director = HouseDirector(builder: builder)
let house = director.construct()
print(house.description)

let builder2 = WoodenHouseBuilder()
let director2 = HouseDirector(builder: builder2)
let house2 = director2.construct()
print(house2.description)

/*
House of Concrete Wall and Concrete Roof
House of Wooden Wall and Wooden Roof
*/
